Add documentation for OTel top-level spans identification (#22277)
DataDog/documentation: content/en/opentelemetry/schema_semantics/service_entry_spans.md

---
title: Mapping OpenTelemetry Semantic Conventions to Service-entry Spans
kind: documentation
aliases:
- /opentelemetry/guide/service_entry_spans_mapping/
further_reading:
- link: "/opentelemetry/integrations/trace_metrics"
  tag: "Documentation"
  text: "OpenTelemetry Trace Metrics"
---

<div class="alert alert-info">
This feature is in beta. If you have any feedback, contact <a href="/help/">Datadog support</a>.
</div>

## Overview

> Datadog uses [service-entry spans][1] throughout the platform for features such as [Trace Metrics][2] and the [APM Trace Explorer][3]. This convention is unique to Datadog, but can be mapped from the [`SpanKind`][4] attribute in OpenTelemetry.

Reviewer: ```suggestion
Datadog uses [service-entry spans][1] throughout the platform for features such as [Trace Metrics][2] and the [APM Trace Explorer][3]. This convention is unique to Datadog, but can be mapped from the [`SpanKind`][4] attribute in OpenTelemetry by following the opt-in guide below.
```

## Opting in to the feature

> To opt into the **public beta**, enable the feature flag depending on the ingestion path.

Reviewer: ```suggestion
To opt into the **public beta**, enable the config option depending on the ingestion path.
```

{{< tabs >}}
{{% tab "OTel Collector and Datadog Exporter" %}}

> The new service-entry span identification logic can be enabled by setting the [`traces::compute_top_level_by_span_kind`][1] config option to true in the Datadog exporter and Datadog connector. This config option needs to be enabled in both the exporter and connector if both components are being used.

Reviewer: ```suggestion
The new service-entry span identification logic can be enabled by setting the `traces::compute_top_level_by_span_kind` config option to true in the [Datadog exporter][2] and [Datadog connector][1]. This config option needs to be enabled in both the exporter and connector if both components are being used.
```

> [1]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.100.0/connector/datadogconnector/examples/config.yaml#L48-L53

Reviewer: Add a link to Datadog exporter too?

{{< /tabs >}}

## Supported conventions

[Trace Metrics][2] are generated for service entry spans and measured spans. These span conventions are unique to Datadog, so OpenTelemetry spans are identified with the following mapping:
| OpenTelemetry Convention | Datadog Convention |
| --- | --- |
| Root span | Service entry span |
| Server span (`span.kind: server`) | Service entry span |
| Consumer span (`span.kind: consumer`) | Service entry span |
| Client span (`span.kind: client`) | Measured span |
| Producer span (`span.kind: producer`) | Measured span |
| Internal span (`span.kind: internal`) | No trace metrics generated |

## Migration

This new service-entry span identification logic may increase the number of spans that generate trace metrics, which may affect existing monitors that are based on trace metrics. Users who only have internal spans will see a decrease in trace metrics.

If you have existing monitors based on trace metrics, you can update them after upgrading since this change introduces more consistency in trace metrics. If you only have internal span, update your instrumentation according to the above table to receive trace metrics and service-entry spans.

> [`SpanKind`][1] is typically set when a span is created, but can also be updated by using the [transform processor][5] in the OpenTelemetry Collector to control the mapping above. For example, if trace metrics are desired for an internal span, the following configuration transforms an internal span with `http.path: "/health"` into a client span:

Reviewer: ```suggestion
[`SpanKind`][4] is typically set when a span is created, but can also be updated by using the [transform processor][5] in the OpenTelemetry Collector to control the mapping above. For example, if trace metrics are desired for an internal span, the following configuration transforms an internal span with `http.path: "/health"` into a client span:
```
